node-js-basics-101/node_modules/ws/lib/Receiver.js

```diff
@@ -0,0 +1,555 @@
+
+      data = this.readBuffer(this.payloadLength);
+      if (this.masked) bufferUtil.unmask(data, this.mask);
+    }
+
+    if (this.opcode > 0x07) {
+      this.controlMessage(data);
+    } else if (this.compressed) {
+      this.state = INFLATING;
+      this.decompress(data);
+    } else if (this.pushFragment(data)) {
+      this.dataMessage();
+    }
+  }
+
+  /**
+   * Decompresses data.
+   *
+   * @param {Buffer} data Compressed data
+   * @private
+   */
+  decompress (data) {
+    const extension = this.extensions[PerMessageDeflate.extensionName];
+
+    extension.decompress(data, this.fin, (err, buf) => {
+      if (err) {
+        this.error(err, err.closeCode === 1009 ? 1009 : 1007);
+        return;
+      }
+
+      if (this.pushFragment(buf)) this.dataMessage();
+      this.startLoop();
+    });
+  }
+
+  /**
+   * Handles a data message.
+   *
+   * @private
+   */
+  dataMessage () {
+    if (this.fin) {
+      const messageLength = this.messageLength;
+      const fragments = this.fragments;
+
+      this.totalPayloadLength = 0;
+      this.messageLength = 0;
+      this.fragmented = 0;
+      this.fragments = [];
+
+      if (this.opcode === 2) {
+        var data;
+
+        if (this.binaryType === 'nodebuffer') {
+          data = toBuffer(fragments, messageLength);
+        } else if (this.binaryType === 'arraybuffer') {
+          data = toArrayBuffer(toBuffer(fragments, messageLength));
+        } else {
+          data = fragments;
+        }
+
+        this.onmessage(data, { masked: this.masked, binary: true });
+      } else {
+        const buf = toBuffer(fragments, messageLength);
+
+        if (!isValidUTF8(buf)) {
+          this.error(new Error('invalid utf8 sequence'), 1007);
+          return;
+        }
+
+        this.onmessage(buf.toString(), { masked: this.masked });
+      }
+    }
+
+    this.state = GET_INFO;
+  }
+
+  /**
+   * Handles a control message.
+   *
+   * @param {Buffer} data Data to handle
+   * @private
+   */
+  controlMessage (data) {
+    if (this.opcode === 0x08) {
+      if (data.length === 0) {
+        this.onclose(1000, '', { masked: this.masked });
+        this.loop = false;
+        this.cleanup(this.cleanupCallback);
+      } else if (data.length === 1) {
+        this.error(new Error('invalid payload length'), 1002);
+      } else {
+        const code = data.readUInt16BE(0, true);
+
+        if (!ErrorCodes.isValidErrorCode(code)) {
+          this.error(new Error(`invalid status code: ${code}`), 1002);
+          return;
+        }
+
+        const buf = data.slice(2);
+
+        if (!isValidUTF8(buf)) {
+          this.error(new Error('invalid utf8 sequence'), 1007);
+          return;
+        }
+
+        this.onclose(code, buf.toString(), { masked: this.masked });
+        this.loop = false;
+        this.cleanup(this.cleanupCallback);
+      }
+
+      return;
+    }
+
+    const flags = { masked: this.masked, binary: true };
```

Comment:
__[jshint]__
* 'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).

<!-- Comment by [SideCI](https://sideci.com) -->